packages/salesforcedx-vscode-apex-testing/src/utils/testUtils.ts

```diff
@@ -0,0 +1,184 @@
+ * Fetch tests from the Tooling API Test Discovery endpoint
+ */
+const fetchFromApi = async (options?: {
+  namespacePrefix?: string;
+}): Promise<{ tests: ApexTestMethod[]; durationMs: number }> => {
+  const start = Date.now();
+  // API path already emits its own Start/End events with source=api internally
+  const result = await discoverTests({
+    showAllMethods: true,
+    namespacePrefix: options?.namespacePrefix
+  });
+  const tests = await convertApiToApexTestMethods(result.classes ?? []);
+  const durationMs = Date.now() - start;
+  return { tests, durationMs };
+};
+
+/**
+ * Returns Apex tests using the configured discovery source.
+ * - ls: queries the Language Server via Apex extension
+ * - api: queries the Tooling API Test Discovery endpoint
+ * Also emits timing metrics and telemetry.
+ */
+export const getApexTests = async (): Promise<ApexTestMethod[]> => {
+  const config = vscode.workspace.getConfiguration('salesforcedx-vscode-apex');
+  const source = config.get<'ls' | 'api'>('testing.discoverySource', 'ls');
+
+  // Fetch according to user selection
+  const selected = source === 'ls' ? await fetchFromLs() : await fetchFromApi();
+  return selected.tests;
+};
+
+/**
+ * Get language client status from Apex extension
+ */
+export const getLanguageClientStatus = async (): Promise<{
+  isReady: () => boolean;
+  isIndexing: () => boolean;
+  failedToInitialize: () => boolean;
+  getStatusMessage: () => string;
+}> => {
+  try {
+    const apexExtension = await getApexExtension();
+    if (!apexExtension?.isActive) {
+      return {
+        isReady: () => false,
+        isIndexing: () => false,
+        failedToInitialize: () => true,
+        getStatusMessage: () => 'Apex extension is not active'
+      };
+    }
+    const apexApi = apexExtension.exports;
+    if (!apexApi?.languageClientManager) {
+      return {
+        isReady: () => false,
+        isIndexing: () => false,
+        failedToInitialize: () => true,
+        getStatusMessage: () => 'Apex extension API is not available'
+      };
+    }
+    return apexApi.languageClientManager.getStatus();
+  } catch (error) {
+    console.debug('Failed to get language client status:', error);
+    return {
+      isReady: () => false,
+      isIndexing: () => false,
+      failedToInitialize: () => true,
+      getStatusMessage: () => 'Failed to get language client status'
+    };
+  }
+};
+
+const buildMeasuresFromTests = (tests: ApexTestMethod[], durationMs: number) => {
+  const numClasses = new Set(tests.map(t => t.definingType)).size;
+  const numMethods = tests.length;
+  return { durationMs, numClasses, numMethods };
+};
+
+/**
+ * Convert API test discovery results to ApexTestMethod format with file locations.
+ * Note: The Tooling API does not provide line/column info, so all locations point to the start of the file (0,0).
+ * For precise method locations, the Language Server discovery should be used instead.
+ */
+const convertApiToApexTestMethods = async (
```

Comment:
@CristiCanizales are there any plan to use this extension w/o an apex language server present, either old or new?

If not, then why not use the language server to get the position data using document symbol request. 

```
const symbols = await vscode.commands.executeCommand(
  'vscode.executeDocumentSymbolProvider',
  uri
);
```